pkg/storage/gcp/bucket.go

```diff
@@ -0,0 +1,21 @@
+package gcp
+
+import (
+	"context"
+	"io"
+)
+
+// Bucket provides file operations for a Google Cloud Storage resource.
+type Bucket interface {
+	// Delete removes the file module/@v/version.extension
+	Delete(ctx context.Context, module, version string) error
+	// Open returns a reader for module/@v/version.extension and any error
+	Open(ctx context.Context, module, version, extension string) (io.ReadCloser, error)
+	// Write returns a new writer for module/@v/version.extension
+	// This writer will overwrite any existing file stored at the same path
+	Write(ctx context.Context, module, version, extension string) io.WriteCloser
+	// ListVersions returns a slice of versions for a module and any error
+	ListVersions(ctx context.Context, module string) ([]string, error)
+	// Exists returns true if the module @ version exists
+	Exists(ctx context.Context, module, version string) bool
+}
```

Comment:
I think it would be more elegant if we could decouple the `Bucket` from our domain. i.e. the Bucket doesn't have to know anything about `module, version, extension` it knows only `path`. i.e.
`Open(ctx context.Context, path string) (io.ReadCloser, error)`

This would make the parallel uploads much easier to implement because I could handle uploading of the different parts (info, mod, zip) inside of the Upload func.